File: api_gateway/src/ui/controller/purchase/order_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/purchase"
	service "github.com/serp/api-gateway/src/core/service/purchase"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type OrderController struct {
	orderService service.IOrderService
}

func (o *OrderController) CreateOrder(ctx *gin.Context) {
	var req request.CreateOrderRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.CreateOrder(ctx.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) UpdateOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.UpdateOrderRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.UpdateOrder(ctx.Request.Context(), orderId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) DeleteOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.DeleteOrder(ctx.Request.Context(), orderId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) GetOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.GetOrder(ctx.Request.Context(), orderId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) GetOrders(ctx *gin.Context) {
	page, size, valid := utils.ValidatePaginationParams(ctx)
	if !valid {
		return
	}

	sortBy := utils.ParseStringQuery(ctx, "sortBy")
	sortDirection := utils.ParseStringQuery(ctx, "sortDirection")
	query := utils.ParseStringQuery(ctx, "query")
	statusId := utils.ParseStringQuery(ctx, "statusId")
	saleChannelId := utils.ParseStringQuery(ctx, "saleChannelId")
	fromSupplierId := utils.ParseStringQuery(ctx, "fromSupplierId")
	orderDateAfter := utils.ParseStringQuery(ctx, "orderDateAfter")
	orderDateBefore := utils.ParseStringQuery(ctx, "orderDateBefore")
	deliveryBefore := utils.ParseStringQuery(ctx, "deliveryBefore")
	deliveryAfter := utils.ParseStringQuery(ctx, "deliveryAfter")

	params := &request.GetOrderParams{
		Page:            &page,
		Size:            &size,
		SortBy:          sortBy,
		SortDirection:   sortDirection,
		Query:           query,
		StatusId:        statusId,
		SaleChannelId:   saleChannelId,
		FromSupplierId:  fromSupplierId,
		OrderDateAfter:  orderDateAfter,
		OrderDateBefore: orderDateBefore,
		DeliveryBefore:  deliveryBefore,
		DeliveryAfter:   deliveryAfter,
	}

	res, err := o.orderService.GetOrders(ctx.Request.Context(), params)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) AddProductToOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.AddOrderItemRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.AddProductToOrder(ctx.Request.Context(), orderId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) DeleteProductFromOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	orderItemId := ctx.Param("orderItemId")
	if orderId == "" || orderItemId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.DeleteProductFromOrder(ctx.Request.Context(), orderId, orderItemId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) UpdateProductInOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	orderItemId := ctx.Param("orderItemId")
	if orderId == "" || orderItemId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.UpdateOrderItemRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.UpdateProductInOrder(ctx.Request.Context(), orderId, orderItemId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) ApproveOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.ApproveOrder(ctx.Request.Context(), orderId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) CancelOrder(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.CancelOrderRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.CancelOrder(ctx.Request.Context(), orderId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (o *OrderController) MarkOrderAsReady(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := o.orderService.MarkOrderAsReady(ctx.Request.Context(), orderId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func NewOrderController(orderService service.IOrderService) *OrderController {
	return &OrderController{
		orderService: orderService,
	}
}

File: purchase_service/src/main/java/serp/project/purchase_service/controller/OrderController.java
package serp.project.purchase_service.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import serp.project.purchase_service.dto.request.OrderCancellationForm;
import serp.project.purchase_service.dto.request.OrderCreationForm;
import serp.project.purchase_service.dto.request.OrderItemUpdateForm;
import serp.project.purchase_service.dto.request.OrderUpdateForm;
import serp.project.purchase_service.dto.response.GeneralResponse;
import serp.project.purchase_service.dto.response.OrderDetailResponse;
import serp.project.purchase_service.dto.response.PageResponse;
import serp.project.purchase_service.entity.OrderEntity;
import serp.project.purchase_service.exception.AppErrorCode;
import serp.project.purchase_service.exception.AppException;
import serp.project.purchase_service.service.OrderItemService;
import serp.project.purchase_service.service.OrderService;
import serp.project.purchase_service.util.AuthUtils;

import java.time.LocalDate;

@RestController
@RequiredArgsConstructor
@RequestMapping("/purchase-service/api/v1/order")
@Validated
@Slf4j
public class OrderController {

        private final OrderService orderService;
        private final OrderItemService orderItemService;
        private final AuthUtils authUtils;

        @PostMapping("/create")
        public ResponseEntity<GeneralResponse<?>> createOrder(@RequestBody OrderCreationForm form) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                Long userId = authUtils.getCurrentUserId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Create order {} by userId {} and tenantId {}",
                                form.getOrderName(), userId, tenantId);
                orderService.createOrder(form, userId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Order created successfully"));
        }

        @PatchMapping("/update/{orderId}")
        public ResponseEntity<GeneralResponse<?>> updateOrder(
                        @RequestBody OrderUpdateForm form,
                        @PathVariable String orderId) {

                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Update order {} for tenantId {}", orderId, tenantId);
                orderService.updateOrder(orderId, form, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Order updated successfully"));
        }

        @PostMapping("/create/{orderId}/add")
        public ResponseEntity<GeneralResponse<?>> addProductToOrder(
                        @RequestBody OrderCreationForm.OrderItem itemForm,
                        @PathVariable String orderId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Add product ID {} to order ID {} for tenantId {}", itemForm.getProductId(),
                                orderId, tenantId);
                orderItemService.createOrderItems(itemForm, orderId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Product added to order successfully"));
        }

        @PatchMapping("/update/{orderId}/delete/{orderItemId}")
        public ResponseEntity<GeneralResponse<?>> deleteProductFromOrder(
                        @PathVariable String orderId,
                        @PathVariable String orderItemId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Delete order item ID {} from order ID {} for tenantId {}", orderItemId,
                                orderId, tenantId);
                orderItemService.deleteOrderItem(orderItemId, orderId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Product removed from order successfully"));
        }

        @PatchMapping("/update/{orderId}/update/{orderItemId}")
        public ResponseEntity<GeneralResponse<?>> updateProductInOrder(
                        @RequestBody OrderItemUpdateForm itemForm,
                        @PathVariable String orderId,
                        @PathVariable String orderItemId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Update order item ID {} in order ID {} for tenantId {}", orderItemId,
                                orderId, tenantId);
                orderItemService.updateOrderItem(orderItemId, itemForm, orderId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Product updated in order successfully"));
        }

        @DeleteMapping("/delete/{orderId}")
        public ResponseEntity<GeneralResponse<?>> deleteOrder(
                        @PathVariable String orderId) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PatchMapping("/manage/{orderId}/approve")
        public ResponseEntity<GeneralResponse<?>> approveOrder(
                        @PathVariable String orderId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                Long userId = authUtils.getCurrentUserId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Approve order {} by userId {} and tenantId {}", orderId, userId, tenantId);
                orderService.approveOrder(orderId, userId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Order approved successfully"));
        }

        @PatchMapping("/manage/{orderId}/cancel")
        public ResponseEntity<GeneralResponse<?>> cancelOrder(
                        @PathVariable String orderId,
                        @RequestBody OrderCancellationForm form) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                Long userId = authUtils.getCurrentUserId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Cancel order {} by userId {} and tenantId {}", orderId, userId, tenantId);
                orderService.cancelOrder(orderId, form.getNote(), userId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Order cancelled successfully"));
        }

        @PatchMapping("/update/{orderId}/ready")
        public ResponseEntity<GeneralResponse<?>> markOrderAsReady(
                        @PathVariable String orderId) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @GetMapping("/search/{orderId}")
        public ResponseEntity<GeneralResponse<OrderDetailResponse>> getOrderDetail(
                        @PathVariable String orderId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                var order = orderService.getOrder(orderId, tenantId);
                if (order == null) {
                        throw new AppException(AppErrorCode.NOT_FOUND);
                }
                log.info("[OrderController] Get order detail for order ID {} and tenantId {}", orderId, tenantId);
                var orderItems = orderItemService.findByOrderId(orderId, tenantId);
                OrderDetailResponse response = OrderDetailResponse.fromEntity(
                                order,
                                orderItems);
                return ResponseEntity.ok(GeneralResponse.success("Successfully get order detail", response));
        }

        @GetMapping("/search")
        public ResponseEntity<GeneralResponse<PageResponse<OrderEntity>>> getOrders(
                        @RequestParam(required = false, defaultValue = "1") int page,
                        @RequestParam(required = false, defaultValue = "10") int size,
                        @RequestParam(required = false, defaultValue = "createdStamp") String sortBy,
                        @RequestParam(required = false, defaultValue = "desc") String sortDirection,
                        @RequestParam(required = false) String query,
                        @RequestParam(required = false) String statusId,
                        @RequestParam(required = false) String fromSupplierId,
                        @RequestParam(required = false) String saleChannelId,
                        @RequestParam(required = false) LocalDate orderDateAfter,
                        @RequestParam(required = false) LocalDate orderDateBefore,
                        @RequestParam(required = false) LocalDate deliveryBefore,
                        @RequestParam(required = false) LocalDate deliveryAfter) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[OrderController] Search orders of page {}/{} for tenantId {}", page, size, tenantId);
                var orders = orderService.findOrders(
                                query,
                                fromSupplierId,
                                saleChannelId,
                                orderDateAfter,
                                orderDateBefore,
                                deliveryBefore,
                                deliveryAfter,
                                statusId,
                                tenantId,
                                page,
                                size,
                                sortBy,
                                sortDirection);
                return ResponseEntity.ok(GeneralResponse.success("Successfully get orders", PageResponse.of(orders)));
        }

}
